### PR TITLE
[workers-shared] Add support for inline comments in _redirects files

### LINKED ISSUE #3255: 🚀 Feature Request: Allow inline comments on redirects
### Describe the solution

In `_redirects`, the following should be valid:

```
/hello /world # a comment
```

### PR DESCRIPTION
**Pull request by petebacondarwin**
Fixes #3255.

Adds support for inline comments in `_redirects` files using the `#` character. This allows developers to document complex redirect rules directly alongside the rules themselves:

```
/old-page /new-page 301 # Moved during site redesign
/blog/* /articles/:splat # Blog URL migration
```

The implementation detects inline comments by looking for any token starting with `#` at position 2 or later (after the required `from` and `to` tokens). This approach:
- Correctly preserves URL fragments like `/page#section` (no space before the `#`)
- Strips comments like `/a /b # comment` or `/a /b 301 # comment`
- Is backwards compatible - existing `_redirects` files work identically
- Handles edge cases like empty comments, multiple hashes in fragments, and absolute URLs with fragments

---

- Tests
  - [x] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [ ] Additional testing not necessary because:
- Public documentation
  - [x] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/28171
  - [ ] Documentation not necessary because: 

*A picture of a cute animal (not mandatory, but encouraged)*

---

<a href="https://app.devin.ai/review/cloudflare/workers-sdk/pull/12467" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by petebacondarwin** [comment]
The [docs](https://developers.cloudflare.com/workers/static-assets/redirects/#per-line) already state:

> Lines starting with a # will be treated as comments.

I think we probably need to tweak those to also cover these new inline comments.

**Comment by petebacondarwin** [comment]
The change is in workers-shared, which is used in the backend to parse the _redirects file at deploy/runtime.
Once we land this change it won't be live in production until the backend updates its version of workers-shared.

File: packages/workers-shared/utils/configuration/parseRedirects.ts [hunk 1]
```diff
@@ -53,7 +53,16 @@ export function parseRedirects(
...
+		let tokens = line.split(/\s+/);
+
+		// Handle inline comments: truncate at first `#` token that starts a comment
+		// This allows `/a /b#fragment` but strips `/a /b # comment`
+		const commentIndex = tokens.findIndex(
+			(token, index) => index >= 2 && token.startsWith("#")
+		);
+		if (commentIndex !== -1) {
+			tokens = tokens.slice(0, commentIndex);
+		}
```
Comment (vicb) [inline packages/workers-shared/utils/configuration/parseRedirects.ts]: I believe this could be simplified with a simple regex: `/(?:^|s+)#.*$/`

Comment (petebacondarwin) [reply packages/workers-shared/utils/configuration/parseRedirects.ts]: I'm surprised that you blocked the PR for this. Feels more like a nit.

The logic is correct but you feel it could be defined in a different way. 
Can you confirm exactly how this would work? It seems to me we would need to apply the regex on the `line` before splitting into `tokens`?

Comment (vicb) [inline packages/workers-shared/utils/configuration/parseRedirects.ts]: Something like:

```suggestion
		// Split tokens by whitespace ignoring comments
		// Note: this allows `/a /b#fragment` but strips `/a /b # comment`
		const tokens = line.replace(/(?:^|\s+)#.*$/, "").split(/\s+/);
```

File: packages/workers-shared/utils/tests/parseRedirects.valid.test.ts [hunk 2]
```diff
@@ -209,6 +262,89 @@ test("parseRedirects should accept relative URLs that don't point to .html files
...
+test("parseRedirects should support inline comments after URL fragments", () => {
+	const input = `/a /b#section 301 # redirect to section`;
+	const result = parseRedirects(input);
+	expect(result).toEqual({
+		rules: [{ from: "/a", status: 301, to: "/b#section", lineNumber: 1 }],
+		invalid: [],
+	});
+});
+
+test("parseRedirects should support inline comments without space before hash", () => {
```
Comment (vicb) [inline packages/workers-shared/utils/tests/parseRedirects.valid.test.ts]:
```suggestion
test("parseRedirects should support inline comments without space after hash", () => {
```

File: packages/workers-shared/utils/configuration/parseRedirects.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/workers-shared/utils/configuration/parseRedirects.ts]: LOL! I literally just pushed that.

File: packages/workers-shared/utils/configuration/parseRedirects.ts [hunk 3]
```diff
@@ -53,7 +53,9 @@ export function parseRedirects(
 			continue;
 		}
 
-		const tokens = line.split(/\s+/);
+		// Handle inline comments: strip off first `#` token that starts a comment but not a URL fragment.
```
Comment (vicb) [inline packages/workers-shared/utils/configuration/parseRedirects.ts]: This comment is wrong but you can use what I proposed earlier

File: packages/workers-shared/utils/configuration/parseRedirects.ts [hunk 4]
```diff
@@ -53,7 +53,7 @@ export function parseRedirects(
 			continue;
 		}
 
-		// Handle inline comments: strip off first `#` token that starts a comment but not a URL fragment.
+		// Handle inline comments: strip off from the first `#` token that starts a comment, but not URL fragments.
```
Comment (vicb) [inline packages/workers-shared/utils/configuration/parseRedirects.ts:56]: What about "# comment" ?

Comment (petebacondarwin) [reply packages/workers-shared/utils/configuration/parseRedirects.ts:56]: that is already handled by previous code in this function

Comment (petebacondarwin) [reply packages/workers-shared/utils/configuration/parseRedirects.ts:56]: see line 43

Comment (vicb) [reply packages/workers-shared/utils/configuration/parseRedirects.ts:56]: Oh right, I missed this.

I think the best way to handle this would be to move all comments handling l43:

```
const line = (lines[i] || "").trim().replace(/(?:^|\s+)#.*$/, "");
```

Handling comments at different places is error prone and might also lead to endless discussion on PRs ;)

Comment (vicb) [reply packages/workers-shared/utils/configuration/parseRedirects.ts:56]: An other option could be to apply the regexp on the input with the `m` flag.

Comment (petebacondarwin) [reply packages/workers-shared/utils/configuration/parseRedirects.ts:56]: The code is only a few lines away, and probably only got missed because that bit of code was not in the diff as it existed prior to this PR.
